src/loader/gtfs/shape_prepare.cc

```diff
@@ -67,71 +72,237 @@ std::vector<shape_offset_t> get_offsets_by_dist_traveled(
+    if (std::all_of(
+#if __cpp_lib_execution
+            std::execution::par_unseq,
+#endif
+            results.begin(), results.end(),
+            [&](shape_results::result const& it) {
+              return *it.stop_seq_ != trip.stop_seq_;
+            })) {
+      auto const distances =
+          trip.distance_traveled_.empty() ? nullptr : &trip.distance_traveled_;
+      results.emplace_back(&trip.stop_seq_, std::move(distances));
+    }
+  }
+}
+
+void shape_prepare::calculate_results(timetable const& tt,
+                                      shape_loader_state const& states) {
+  auto m = std::mutex{};
   auto const progress_tracker = utl::get_active_progress_tracker();
   progress_tracker->status("Calculating shape offsets")
-      .out_bounds(98.F, 100.F)
-      .in_high(trips.size());
+      .out_bounds(98.F, 99.F)
+      .in_high(shape_results_.size());
+  std::for_each(
+#if __cpp_lib_execution
+      std::execution::par_unseq,
+#endif
+      shape_results_.begin(), shape_results_.end(),
+      [&](shape_results& results) {
+        progress_tracker->increment();
+        if (results.results_.empty()) {
+          return;
+        }
+        auto const shape = shapes_.get_shape(results.shape_idx_);
+        auto const& shape_distances =
+            states
+                .distances_[cista::to_idx(results.shape_idx_ - index_offset_)];
+        for (auto& result : results.results_) {
+          auto const offsets = [&]() {
+            if (!shape_distances.empty() && result.distances_ != nullptr) {
+              return get_offsets_by_dist_traveled(*result.distances_,
+                                                  shape_distances);
+            }
+            if (shape.size() < result.stop_seq_->size()) {
+              return std::vector<shape_offset_t>{};
+            }
 
-  auto const key_hash =
-      [](std::pair<shape_idx_t, stop_seq_t const*> const& pair) noexcept {
-        auto h = cista::BASE_HASH;
-        h = cista::hash_combine(h, cista::hashing<shape_idx_t>{}(pair.first));
-        h = cista::hash_combine(h, cista::hashing<stop_seq_t>{}(*pair.second));
-        return h;
-      };
-  auto const key_compare =
-      [](std::pair<shape_idx_t, stop_seq_t const*> const& a,
-         std::pair<shape_idx_t, stop_seq_t const*> const& b) noexcept {
-        return (a.first == b.first) && (*a.second == *b.second);
-      };
-  auto shape_offsets_cache =
-      hash_map<std::pair<shape_idx_t, stop_seq_t const*>, shape_offset_idx_t,
-               decltype(key_hash), decltype(key_compare)>{};
+            return get_offsets_by_stops(tt, shape, *result.stop_seq_);
+          }();
+          if (!offsets.empty()) {
+            auto const guard = std::lock_guard<decltype(m)>{m};
+            result.offset_idx_ = shapes_.add_offsets(offsets);
+          } else {
+            result.offset_idx_ = shape_offset_idx_t::invalid();
+          }
+          result.boxes_ = [&]() {
+            // Store box of full shape at index 0
```

Comment:
Nicht offensichtlich an anderen Codestellen.
Vielleicht besser `vector_map<route_idx_t, box> route_bbox_` + `vecvec<...> route_segment_bboxes_`?

Reply:
Should be changed now. I forgot to update variable names.